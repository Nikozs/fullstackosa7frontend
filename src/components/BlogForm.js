import React from "react";
import { Form , Button, Container } from "react-bootstrap";
import { blogCreation } from "./../reducers/blogsReducer";
import { newblogNotification, thunkNotify, clearNotification } from "./../reducers/notificationReducer";
import { connect } from "react-redux";

class BlogForm extends React.Component {

	addBlogi = async (event) => {
		event.preventDefault();
		const blogObjecti = {
			title: event.target.newTitle.value,
			author: event.target.newAuthor.value,
			url: event.target.newUrl.value,
			likes: 0,
			user: this.props.user
		};

		await	this.props.blogCreation(blogObjecti);
		this.props.thunkNotify(`Added blog: '${blogObjecti.title}'`, 10000);
	}

handleBlogChange = (event) => {
	this.setState({ [event.target.name]: event.target.value });
}


render() {
	return (
		<div>
			<Container>
				<h2>Create a new blog</h2>

				<Form onSubmit={this.addBlogi}>
					<Form.Label>Title: </Form.Label>
					<Form.Control
						name="newTitle"
						value={this.newTitle}
						onChange={this.handleBlogChange}
					/>
					<Form.Label>Blog text:</Form.Label>
					<Form.Control
						name="newBlog"
						value={this.newBlog}
						onChange={this.handleBlogChange}
					/>
					<Form.Label>Author:</Form.Label>
					<Form.Control
						name="newAuthor"
						value={this.newAuthor}
						onChange={this.handleBlogChange}
					/>
					<Form.Label>URL-link:</Form.Label>
					<Form.Control
						name="newUrl"
						value={this.newUrl}
						onChange={this.handleBlogChange}
					/>
					<br />

					<Button variant="primary" type="submit">Save</Button>
				</Form>
			</Container>
		</div>
	);
}
}


const mapStateToProps = (state) => {
	return {
		newTitle: state.newTitle,
		newBlog: state.newBlog,
		newAuthor: state.newAuthor,
		newUrl: state.newUrl,
		likes: state.newLikes,
		user: state.logins.user
	};
};

export default connect(
	mapStateToProps,
	{ blogCreation, thunkNotify, newblogNotification, clearNotification }
)(BlogForm);



