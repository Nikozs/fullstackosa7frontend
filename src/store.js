import { combineReducers, createStore, applyMiddleware } from "redux";
import notificationReducer from "./reducers/notificationReducer";
import userReducer from "./reducers/usersReducer";
import blogsReducer from "./reducers/blogsReducer";
import loginReducer from "./reducers/loginReducer";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import commentsReducer from "./reducers/commentsReducer";


const persistConfig = {
	key: "root",
	storage
};

const reducer = combineReducers({
	users: userReducer,
	notification: notificationReducer,
	blogs: blogsReducer,
	comments: commentsReducer,
	logins: loginReducer
});

const persistedReducer = persistReducer(persistConfig, reducer);

export const store = createStore(
	persistedReducer,
	composeWithDevTools(
		applyMiddleware(thunk)
	)
);

export const persistor = persistStore(store);

