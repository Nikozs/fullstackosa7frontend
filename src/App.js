import React from "react";
import Blog from "./components/Blog";
import User from "./components/User";
import blogService from "./services/blogs";
import Notification from "./components/Notification";
import Togglable from "./components/Togglable";
import LoginForm from "./components/Loginform";
import UsersList from "./components/Users";
import { BrowserRouter as Router, Route, Link, Switch, NavLink } from "react-router-dom";
import {  Form, Button, } from "react-bootstrap";
import { getBlog } from "./reducers/blogsReducer";
import { blogInitialization } from "./reducers/blogsReducer";
import { usersInitialization } from "./reducers/usersReducer";
import { commentInitialization } from "./reducers/commentsReducer";
import { Logout } from "./reducers/loginReducer";
import { connect } from "react-redux";
import BlogList from "./components/BlogList";
import BlogForm from "./components/BlogForm";
import { persistor } from "./store";

class App extends React.Component {

	componentDidMount() {
		this.props.usersInitialization();
		this.props.blogInitialization();
		this.props.commentInitialization();
		try {
			if ( this.props.user !== null )
			{

				blogService.setToken(this.props.user.token);

			}
		}
		catch (e) {
			// statements to handle any exceptions
			//logMyErrors(e); // pass exception object to error handler
		}


	}


	handleBlogChange = (event) => {
		this.setState({ [event.target.name]: event.target.value });
	}


	toggleVisible = () => {
		this.setState({ showAll: !this.state.showAll });
	}

	handleLoginFieldChange = (event) => {
		this.setState({ [event.target.name]: event.target.value });
	}


	logout = (event) => {
		event.preventDefault();

		this.props.Logout();
		persistor.purge();
		console.log("Kirjauduttu ulos");
	}


	render() {



		const tglloginForm = () => (
			<div>
				<Togglable buttonLabel="Login">
					<LoginForm/>
				</Togglable>

			</div>
		);

		const logoutForm = () => (
			<div>
				<Form onSubmit={this.logout}>
					<Button variant="danger" size="sm" type="submit">Log out</Button>
				</Form>
			</div>
		);


		const navbarStyle = {
			background: "lightgrey"
		};

		const navStyle = {
			fontWeight: "bold",
			color: "black",
			margin: "2px",
			padding: "2px"
		};


		return (

			<div>
				<h1>Blog app</h1>
				<div>
					<Notification />
					<Router>
						<div style={navbarStyle}>

							<NavLink to="/blogs" activeStyle={navStyle} style={{ textDecoration: "none" }}>Blogs</NavLink> &nbsp;
							<NavLink to="/users" activeStyle={navStyle} style={{ textDecoration: "none" }}>Users</NavLink>&nbsp;
							{this.props.user
								? <em>{this.props.username} logged in {this.props.user !== null && logoutForm()}</em>
								: <em>{this.props.user === null && tglloginForm()}</em>
							}&nbsp;

						</div>

						<Link to="/create"><button>Create New</button></Link>

						<Switch>
							<Route exact path="/blogs" render={() => <BlogList />} />
							<Route path="/users" render={() => <UsersList />} />
							<Route path="/create" render={() => <BlogForm blogs={this.props.blogs}/>} />
							<Route exact path="/blog/:id" render={({ match }) =>
								<Blog blog={this.props.blogs.find(blog => blog.id === match.params.id)} />} />
							<Route exact path="/user/:id" render={({ match }) =>
								<User blog={this.props.blogs.find(blog => blog.id === match.params.id)} user={this.props.users.find(user => user.id === match.params.id)} />} />
						</Switch>


					</Router>
				</div>


			</div>
		);
	}
}




const mapStateToProps = (state) => {
	return {
		blogs: state.blogs,
		users: state.users,
		notification: state.notification,
		user: state.logins.user,
		username: state.logins.username
	};
};


export default connect(
	mapStateToProps,
	{ usersInitialization, commentInitialization, blogInitialization, getBlog, Logout }
)(App);