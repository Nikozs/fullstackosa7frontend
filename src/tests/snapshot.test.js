import React from "react";
import { BrowserRouter as Router, Route, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import BlogForm from "./../components/BlogForm";
import { shallow } from "enzyme";
import { Provider , connect } from "react-redux";
import { store, persistor } from "../store";
import { PersistGate } from "redux-persist/integration/react";
import App from "../App";

jest.mock("../services/blogs");




describe("render test", () => {
	it("renders correctly", () => {
		const tree = renderer
			.create(<Router><Link to="http://localhost:3000/">Blog app</Link></Router>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});


describe("BlogForm component", () => {
	it("matches the snapshot", () => {

		const tree = renderer
			.create(<Provider store={store}>
				<PersistGate loading={null} persistor={persistor}><BlogForm />
				</PersistGate>
			</Provider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});