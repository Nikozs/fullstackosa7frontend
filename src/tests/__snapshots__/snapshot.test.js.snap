// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogForm component matches the snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <h2>
      Create a new blog
    </h2>
    <form
      className=""
      onSubmit={[Function]}
    >
      <label
        className="form-label"
      >
        Title: 
      </label>
      <input
        className="form-control"
        name="newTitle"
        onChange={[Function]}
      />
      <label
        className="form-label"
      >
        Blog text:
      </label>
      <input
        className="form-control"
        name="newBlog"
        onChange={[Function]}
      />
      <label
        className="form-label"
      >
        Author:
      </label>
      <input
        className="form-control"
        name="newAuthor"
        onChange={[Function]}
      />
      <label
        className="form-label"
      >
        URL-link:
      </label>
      <input
        className="form-control"
        name="newUrl"
        onChange={[Function]}
      />
      <br />
      <button
        className="btn btn-primary"
        disabled={false}
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;

exports[`render test renders correctly 1`] = `
<a
  href="/http://localhost:3000/"
  onClick={[Function]}
>
  Blog app
</a>
`;
